File: app/src/main/java/com/github/bjoernpetersen/q/ui/fragments/QueueFragment.java
package com.github.bjoernpetersen.q.ui.fragments;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.State;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.github.bjoernpetersen.jmusicbot.client.ApiException;
import com.github.bjoernpetersen.jmusicbot.client.model.QueueEntry;
import com.github.bjoernpetersen.jmusicbot.client.model.Song;
import com.github.bjoernpetersen.q.QueueState;
import com.github.bjoernpetersen.q.R;
import com.github.bjoernpetersen.q.api.Connection;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

/**
 * A fragment representing a list of Items.
 * <p />
 * Activities containing this fragment MUST implement the {@link ListFragmentInteractionListener}
 * interface.
 */
public class QueueFragment extends Fragment {

  private static final String TAG = QueueFragment.class.getSimpleName();
  private static final String ITEMS_KEY = QueueFragment.class.getName() + "items";

  private ArrayList<QueueEntry> items;
  private ListFragmentInteractionListener mListener;
  private QueueState.Listener queueListener;
  private ScheduledExecutorService updater;
  private ScheduledFuture<?> updateTask;

  /**
   * Mandatory empty constructor for the fragment manager to instantiate the
   * fragment (e.g. upon screen orientation changes).
   */
  public QueueFragment() {
  }

  @SuppressWarnings("unused")
  public static QueueFragment newInstance() {
    QueueFragment fragment = new QueueFragment();
    Bundle args = new Bundle();
    fragment.setArguments(args);
    return fragment;
  }

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    if (savedInstanceState != null) {
      this.items = savedInstanceState.getParcelableArrayList(ITEMS_KEY);
    }
    if (this.items == null) {
      this.items = new ArrayList<>();
    }

    updater = Executors.newSingleThreadScheduledExecutor();
  }

  @Override
  public void onDestroy() {
    updater.shutdown();
    updater = null;
    super.onDestroy();
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    View view = inflater.inflate(R.layout.fragment_queue_list, container, false);

    // Set the adapter
    if (view instanceof RecyclerView) {
      Context context = view.getContext();
      RecyclerView recyclerView = (RecyclerView) view;
      recyclerView.setAdapter(new QueueRecyclerViewAdapter(items, mListener));
      recyclerView.addItemDecoration(
          new DividerItemDecoration(recyclerView.getContext(), DividerItemDecoration.VERTICAL)
      );
    } else {
      throw new IllegalStateException();
    }

    return view;
  }

  @Override
  public void onStart() {
    super.onStart();
    updateQueue(QueueState.getInstance().get());
    QueueState.getInstance().addListener(queueListener = new QueueState.Listener() {
      @Override
      public void onChange(List<QueueEntry> oldQueue, List<QueueEntry> newQueue) {
        updateQueue(newQueue);
      }
    });
    updateTask = updater.scheduleWithFixedDelay(new UpdateTask(), 0, 2, TimeUnit.SECONDS);
  }

  @Override
  public void onStop() {
    updateTask.cancel(true);
    QueueState.getInstance().removeListener(queueListener);
    queueListener = null;
    super.onStop();
  }

  @Override
  public void onSaveInstanceState(Bundle outState) {
    outState.putParcelableArrayList(ITEMS_KEY, items);
    super.onSaveInstanceState(outState);
  }

  @Override
  public void onAttach(Context context) {
    super.onAttach(context);
    if (context instanceof ListFragmentInteractionListener) {
      mListener = (ListFragmentInteractionListener) context;
    } else {
      throw new RuntimeException(context.toString()
          + " must implement ListFragmentInteractionListener");
    }
  }

  @Override
  public void onDetach() {
    super.onDetach();
    mListener = null;
  }

  private void updateQueue(List<QueueEntry> queue) {
    items.clear();
    items.addAll(queue);
    items.add(null);
    RecyclerView recyclerView = (RecyclerView) getView();
    if (recyclerView == null) {
      return;
    }
    recyclerView.getAdapter().notifyDataSetChanged();
  }

  private class UpdateTask implements Runnable {

    @Override
    public void run() {
      try {
        final List<QueueEntry> queue = Connection.get(getContext()).getQueue();
        View view = getView();
        if (view == null) {
          return;
        }
        view.post(new Runnable() {
          @Override
          public void run() {
            QueueState.getInstance().set(queue);
          }
        });
      } catch (ApiException e) {
        Log.v(TAG, "Could not get queue", e);
      } catch (RuntimeException e) {
        Log.wtf(TAG, "FUCK", e);
      }
    }
  }

  /**
   * This interface must be implemented by activities that contain this
   * fragment to allow an interaction in this fragment to be communicated
   * to the activity and potentially other fragments contained in that
   * activity.
   * <p/>
   * See the Android Training lesson <a href=
   * "http://developer.android.com/training/basics/fragments/communicating.html"
   * >Communicating with Other Fragments</a> for more information.
   */
  public interface ListFragmentInteractionListener {

    void onSongClick(Song song);

    void showSearch();

    void showSuggestions();
  }
}

File: app/src/main/java/com/github/bjoernpetersen/q/ui/fragments/QueueFragment.kt
package com.github.bjoernpetersen.q.ui.fragments

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.github.bjoernpetersen.jmusicbot.client.ApiException
import com.github.bjoernpetersen.jmusicbot.client.model.QueueEntry
import com.github.bjoernpetersen.q.QueueState
import com.github.bjoernpetersen.q.R
import com.github.bjoernpetersen.q.api.Connection
import com.github.bjoernpetersen.q.api.action.DiscoverHost
import com.github.bjoernpetersen.q.tag
import com.github.bjoernpetersen.q.ui.fragments.QueueEntryAdapter.QueueEntryType
import com.github.bjoernpetersen.q.ui.fragments.QueueEntryAddButtonsDataBinder.QueueEntryAddButtonsListener
import com.github.bjoernpetersen.q.ui.fragments.QueueEntryDataBinder.QueueEntryListener
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.io.IOException
import java.util.concurrent.TimeUnit

private val ITEMS_KEY = QueueFragment::class.java.name + ".items"

/**
 * A fragment representing a list of Items.
 *
 * Activities containing this fragment MUST implement
 * the [QueueEntryListener] and [QueueEntryAddButtonsListener] interface.
 */
class QueueFragment : Fragment() {

  private var dataBinder: QueueEntryDataBinder? = null
  private var addButtonsListener: QueueEntryAddButtonsListener? = null
  private var entryListener: QueueEntryListener? = null
  private var updater: Disposable? = null
  private var listener: ((List<QueueEntry>, List<QueueEntry>) -> Unit)? = null

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
      savedInstanceState: Bundle?): View? {
    val view = inflater.inflate(R.layout.fragment_queue_list, container, false)
        as? RecyclerView ?: throw IllegalStateException()

    val error = { throw IllegalStateException("Should have been set in onAttach!") }
    val adapter = QueueEntryAdapter(entryListener ?: error(), addButtonsListener ?: error())
    view.adapter = adapter
    view.addItemDecoration(DividerItemDecoration(view.context, DividerItemDecoration.VERTICAL))
    val items: List<QueueEntry> =
        savedInstanceState?.getParcelableArrayList(ITEMS_KEY) ?: ArrayList()
    dataBinder = adapter.getDataBinder(QueueEntryType.QUEUE_ENTRY)
    dataBinder!!.items = items
    return view
  }

  override fun onDestroyView() {
    super.onDestroyView()
    dataBinder = null
  }

  override fun onStart() {
    super.onStart()
    updateQueue(QueueState.queue)

    listener = { _, newQueue: List<QueueEntry> -> updateQueue(newQueue) }
    QueueState.addListener(listener!!)

    updater = startUpdater()
  }

  override fun onStop() {
    updater?.dispose()
    updater = null
    QueueState.removeListener(listener!!)
    listener = null
    super.onStop()
  }

  override fun onSaveInstanceState(outState: Bundle?) {
    dataBinder?.apply { outState?.putParcelableArrayList(ITEMS_KEY, ArrayList(items)) }
    super.onSaveInstanceState(outState)
  }

  override fun onAttach(context: Context?) {
    super.onAttach(context)
    entryListener = context as? QueueEntryListener ?:
        throw RuntimeException(context?.toString() + " must implement QueueEntryListener")

    addButtonsListener = context as? QueueEntryAddButtonsListener ?:
        throw RuntimeException(context.toString() + " must implement QueueEntryAddButtonsListener")
  }

  override fun onDetach() {
    super.onDetach()
    entryListener = null
    addButtonsListener = null
  }

  private fun updateQueue(queue: List<QueueEntry>) {
    dataBinder?.items = queue
  }

  private fun startUpdater() = Observable.interval(2, TimeUnit.SECONDS)
      .subscribeOn(Schedulers.io())
      .map { Connection.getQueue() }
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe({
        QueueState.queue = it
      }, {
        Log.d(tag(), "Could not retrieve Queue.")
        when (it) {
          is ApiException -> if (it.cause is IOException) {
            DiscoverHost().defaultAction()
          }
          is RuntimeException -> Log.wtf(tag(), it)
        }
      })

  companion object {
    @JvmStatic
    fun newInstance(): QueueFragment {
      val fragment = QueueFragment()
      val args = Bundle()
      fragment.arguments = args
      return fragment
    }
  }
}
